### ISSUE TITLE
Pulses/Subscription interface loads very slow, when connection to Slack is slow

### ISSUE THREAD

**Issue by flamber** [metabase/metabase#18774]
**Describe the bug**
If Slack is configured and there's slow network to Slack (or having 1000s of channels), then Subscription sidebar is blank until the request finishes. And Pulses interface will fail trying to click the Email/Slack toggles until the (two) request loads.
This is because there's requests to Slack on every request to endpoint `/api/pulse/form_input`

P1 contender.

**To Reproduce**
1. Setup `/api/pulse/form_input`, so it loads slowly. It can be mocked with a reverse-proxy, or [tc](https://wiki.linuxfoundation.org/networking/netem#emulating_wide_area_network_delays) 2 second delay on your WAN interface, when Slack is configured.
2. New tab, go to a dashboard and open the subscription sidebar - it's blank until the request has finished.
Try closing the sidebar and open it again - it will open instantly, but another request is also made

https://user-images.githubusercontent.com/1447303/139534451-3004fa5e-df43-49c2-84f6-3e31a1d94946.mp4

3. New tab, go to :heavy_plus_sign: > New pulse - it's not possible to use the Email/Slack toggles until the request has loaded.
There's also two requests being sent. And if leaving the page and going back, then the toggles work, but another request is made.

https://user-images.githubusercontent.com/1447303/139534470-d6273daf-0f54-400c-a417-1ead05b0b7f4.mp4

The browser console gives the following errors, when trying to use the Pulse Email/Slack toggles before the request is loaded:
```
Uncaught TypeError: Cannot read properties of undefined (reading 'email')
    at i.value (PulseEditChannels.jsx:53)
    at i.value (PulseEditChannels.jsx:113)
    at Toggle.jsx:18
    at Object.u (react-dom.production.min.js:14)
    at p (react-dom.production.min.js:14)
    at react-dom.production.min.js:14
    at v (react-dom.production.min.js:15)
    at ut (react-dom.production.min.js:52)
    at at (react-dom.production.min.js:51)
    at ct (react-dom.production.min.js:52)

Uncaught TypeError: Cannot read properties of undefined (reading 'slack')
    at i.value (PulseEditChannels.jsx:53)
    at i.value (PulseEditChannels.jsx:113)
    at Toggle.jsx:18
    at Object.u (react-dom.production.min.js:14)
    at p (react-dom.production.min.js:14)
    at react-dom.production.min.js:14
    at v (react-dom.production.min.js:15)
    at ut (react-dom.production.min.js:52)
    at at (react-dom.production.min.js:51)
    at ct (react-dom.production.min.js:52)
```

**Expected behavior**
1. Cache the Slack channels in the backend, so it doesn't have to do requests to Slack every time `/api/pulse/form_input` is being accessed. I'm unsure how long they should be cached, or if a flush-cache-endpoint is needed.
Also unsure what happens if there's 1000s of channels and it takes [more than 10 seconds](https://github.com/metabase/metabase/blob/ce939219f8a3e28ecb00deb232154780a1dcfcef/src/metabase/integrations/slack.clj#L58-L59).
2. Subscription sidebar: show loading-spinner and use the cached response instead of making a new request
3. Pulse: show loading-spinner when trying to toggle Email/Slack until request has finished, and only send a single request and use the cached response instead of making a new request

Alternatively, when we change from Bot to App (#12383), then remove the channels from that endpoint completely and have them in a separate searchable endpoint (similar to `/api/user`, which contains the users that can be used for Email).

**Information about your Metabase Installation:**
Tested 0.38.1 thru 0.41.1, but previous versions seem to have similar behavior, though #17479 (fixed in 0.41.0) took care of a race-condition that could occur.

**Comment by escherize** [comment]
> Also unsure what happens if there's 1000s of channels and it takes [more than 10 seconds](https://github.com/metabase/metabase/blob/ce939219f8a3e28ecb00deb232154780a1dcfcef/src/metabase/integrations/slack.clj#L58-L59).

This will not be a problem. The long wait time experienced on the frontend side is spread over multiple calls from the backend to the Slack `conversations.list` API. This api returns a limited number of conversations, and a "cursor" which denotes whether or not there are more conversations to get. Currently, the backend [continues calling the Slack API](https://github.com/metabase/metabase/blob/noahmoss-user-local-settings/src/metabase/integrations/slack.clj#L134) until there are no more, or we get [10,000 results](https://github.com/metabase/metabase/blob/noahmoss-user-local-settings/src/metabase/integrations/slack.clj#L129).

**Comment by escherize** [comment]
I've looked into this a bit more, and figured out why this route can take _so long_. As I stated we can churn through pages and pages of results from the Slack API, and there isn't a good way to paralellize this work. Here's a diagram of our flow: ![](https://mermaid.ink/img/pako:eNqFUs1qHDEMfhXh04ZsS1vIZQ6BDu1AoZTCHHpZKFpbmTE7I09lOaWEvHvt9WaSDQmxL5a_H8my7owNjkxjIv1JxJa-eBwE5x1DXup1oga-B3SeB4gT2gPYEZlpimCRQfFAgDCFDKufqeowaeA070mO8YKi3voFWaEDjNBJYCV2lf0Ubgvc5jQvon1B-2MVn39-q_iPoATih1Eh3EDXwK-RGJjIkduClGdFfVZ6VQpZBRn2uPl0dbX9UPb7jxcV695dX7fNKheKadKTbgphgY7UjqdaUiS5PPcuq80e_erx--SxuXhk4KSPQf-Q8kiDS7BJYpA3_SrtievaumeOGw5wRkar_hZz-9oar8KSqStCTcLnj3f0qurln9CRIOJMay99LJ33dEsOcEDPr6j_FslCqEcLYvVSbhzsp2APZmtmkhm9y8N7Vyx2JvPyBJomHx3dYC56Z3Z8n6lpcbnir85rENOoJNqaMqT9P7YPceWc5r9e3v8HGl0EQQ)

**Comment by escherize** [comment]
I have been working on this, and published some findings [here](https://github.com/metabase/metabase/pull/20570#issuecomment-1049192634). Will hand off the frontend work shortly

**Comment by flamber** [comment]
Fixed by #20570 and #21172, so Slack channels are only updated hourly via a task (or on restart), which should prevent "Too many requests" hitting the Slack API as well as return cached channel names almost instantly to the user in Metabase.
